### ISSUE TITLE
Build step fails

### ISSUE THREAD

**Issue by MatteoGioioso** [aws/copilot-cli#2082]
I have created my first pipeline with all the defaults.

1. `copilot pipeline init`
2. change the trigger branch from `master` to `dev`
3. commit and push the changes
4. build starts
5. after the image has been built and tagged it tries to push it to ECR with: `aws ecr get-login --no-include-email --region $region`
6. Somehow the region is empty and it fails

It fails the build at this particular step (buildspec.yml):

```
for env in $envs; do
            repo=$(cat $CODEBUILD_SRC_DIR/infrastructure/$workload-$env.params.json | jq '.Parameters.ContainerImage' | sed 's/"//g');
            region=$(echo $repo | cut -d'.' -f4); // <- probably empty
            $(aws ecr get-login --no-include-email --region $region);
            docker tag $image_id $repo;
            docker push $repo;
          done;

```

It cannot find the region somehow:

```

aws help
--
872 | aws <command> help
873 | aws <command> <subcommand> help
874 | aws: error: argument --region: expected one argument
875 | "docker tag" requires exactly 2 arguments.
876 | See 'docker tag --help'.

```

Copilot version: `v1.2.0, built for linux`

### Generated buildspec:

<details>
  <summary>Click to expand!</summary>
  
```
# Buildspec runs in the build stage of your pipeline.
version: 0.2
phases:
  install:
    runtime-versions:
      docker: 18
      ruby: 2.6
    commands:
      - echo "cd into $CODEBUILD_SRC_DIR"
      - cd $CODEBUILD_SRC_DIR
      # Download the copilot linux binary.
      - wget https://ecs-cli-v2-release.s3.amazonaws.com/copilot-linux-v1.2.0
      - mv ./copilot-linux-v1.2.0 ./copilot-linux
      - chmod +x ./copilot-linux
  build:
    commands:
      - echo "Perform migration"
#      - cd db/ && copilot task run
  post_build:
    commands:
      - ls -l
      - export COLOR="false"
      # First, upgrade the cloudformation stack of every environment in the pipeline.
      - pipeline=$(cat $CODEBUILD_SRC_DIR/copilot/pipeline.yml | ruby -ryaml -rjson -e 'puts JSON.pretty_generate(YAML.load(ARGF))')
      - pl_envs=$(echo $pipeline | jq '.stages[].name' | sed 's/"//g')
      - >
        for pl_env in $pl_envs; do
        ./copilot-linux env upgrade -n $pl_env;
        done;
      # Find all the local services in the workspace.
      - svcs=$(./copilot-linux svc ls --local --json | jq '.services[].name' | sed 's/"//g')
      # Find all the local jobs in the workspace.
      - jobs=$(./copilot-linux job ls --local --json | jq '.jobs[].name' | sed 's/"//g')
      # Find all the environments
      - envs=$(./copilot-linux env ls --json | jq '.environments[].name' | sed 's/"//g')
      # Generate the cloudformation templates.
      # The tag is the build ID but we replaced the colon ':' with a dash '-'.
      - tag=$(sed 's/:/-/g' <<<"$CODEBUILD_BUILD_ID")
      - >
        for env in $envs; do
          for svc in $svcs; do
          ./copilot-linux svc package -n $svc -e $env --output-dir './infrastructure' --tag $tag;
          done;
          for job in $jobs; do
          ./copilot-linux job package -n $job -e $env --output-dir './infrastructure' --tag $tag;
          done;
        done;
      - ls -lah ./infrastructure
      # Concatenate jobs and services into one var for addons
      # If addons exists, upload addons templates to each S3 bucket and write template URL to template config files.
      - WORKLOADS=$(echo $jobs $svcs)
      - |
        for workload in $WORKLOADS; do
          ADDONSFILE=./infrastructure/$workload.addons.stack.yml
          if [ -f "$ADDONSFILE" ]; then
            tmp=$(mktemp)
            timestamp=$(date +%s)
            aws s3 cp "$ADDONSFILE" "s3://stackset-myapp-infrast-pipelinebuiltartifactbuc-11y4xs1nt414k/manual/$timestamp/$workload.addons.stack.yml";
            jq --arg a "https://stackset-myapp-infrast-pipelinebuiltartifactbuc-11y4xs1nt414k.s3.us-east-1.amazonaws.com/manual/$timestamp/$workload.addons.stack.yml" '.Parameters.AddonsTemplateURL = $a' ./infrastructure/$workload-dev.params.json > "$tmp" && mv "$tmp" ./infrastructure/$workload-dev.params.json
          fi
        done;
      # Build images
      # - For each manifest file:
      #   - Read the path to the Dockerfile by translating the YAML file into JSON.
      #   - Run docker build.
      #   - For each environment:
      #     - Retrieve the ECR repository.
      #     - Login and push the image.
      - >
        for workload in $WORKLOADS; do
          manifest=$(cat $CODEBUILD_SRC_DIR/copilot/$workload/manifest.yml | ruby -ryaml -rjson -e 'puts JSON.pretty_generate(YAML.load(ARGF))')
          image_location=$(echo $manifest | jq '.image.location')
          if [ ! "$image_location" = null ]; then
            echo "skipping image building because location is provided as $image_location";
            continue
          fi
          base_dockerfile=$(echo $manifest | jq '.image.build')
          build_dockerfile=$(echo $manifest| jq 'if .image.build?.dockerfile? then .image.build.dockerfile else "" end' | sed 's/"//g')
          build_context=$(echo $manifest| jq 'if .image.build?.context? then .image.build.context else "" end' | sed 's/"//g')
          build_target=$(echo $manifest| jq 'if .image.build?.target? then .image.build.target else "" end' | sed 's/"//g')
          dockerfile_args=$(echo $manifest | jq 'if .image.build?.args? then .image.build.args else "" end | to_entries?')
          build_cache_from=$(echo $manifest | jq 'if .image.build?.cache_from? then .image.build.cache_from else "" end')
          df_rel_path=$( echo $base_dockerfile | sed 's/"//g')
          if [ -n "$build_dockerfile" ]; then 
            df_rel_path=$build_dockerfile
          fi
          df_path=$df_rel_path
          df_dir_path=$(dirname "$df_path")
          if [ -n "$build_context" ]; then
            df_dir_path=$build_context
          fi
          build_args=
          if [ -n "$dockerfile_args" ]; then
            for arg in $(echo $dockerfile_args | jq -r '.[] | "\(.key)=\(.value)"'); do 
              build_args="$build_args--build-arg $arg "
            done
          fi
          if [ -n "$build_target" ]; then
            build_args="$build_args--target $build_target "
          fi
          if [ -n "$build_cache_from" ]; then
            for arg in $(echo $build_cache_from | jq -r '.[]'); do
              build_args="$build_args--cache-from $arg "
            done
          fi
          echo "Name: $workload"
          echo "Relative Dockerfile path: $df_rel_path"
          echo "Docker build context: $df_dir_path"
          echo "Docker build args: $build_args"
          echo "Running command: docker build -t $workload:$tag $build_args-f $df_path $df_dir_path";
          docker build -t $workload:$tag $build_args-f $df_path $df_dir_path;
          image_id=$(docker images -q $workload:$tag);
          for env in $envs; do
            repo=$(cat $CODEBUILD_SRC_DIR/infrastructure/$workload-$env.params.json | jq '.Parameters.ContainerImage' | sed 's/"//g');
            echo $repo
            region=$(echo $repo | cut -d'.' -f4);
            $(aws ecr get-login --no-include-email --region $region);
            docker tag $image_id $repo;
            docker push $repo;
          done;
        done;
artifacts:
  files:
    - "infrastructure/*"

```
</details>

### Pipeline

<details>
  <summary>Click to expand!</summary>
  
```

name: pipeline-myapp-myservice-service

# The version of the schema used in this template
version: 1

# This section defines the source artifacts.
source:
  # The name of the provider that is used to store the source artifacts.
  provider: GitHub
  # Additional properties that further specifies the exact location
  # the artifacts should be sourced from. For example, the GitHub provider
  # has the following properties: repository, branch.
  properties:
    access_token_secret: github-token-myapp-myservice-service
    branch: dev
    repository: https://github.com/owner/repo

# The deployment section defines the order the pipeline will deploy
# to your environments.
stages:
    - name: dev
      test_commands:
       - make test -C src/

```

</details>

**Comment by huanjani** [comment]
Hello, @MatteoGioioso!

Which version of the aws-cli are you using?

**Comment by bvtujo** [comment]
Hi @MatteoGioioso, can you also post the manifest for the workload on which this step failed? It seems like Copilot is receiving an incorrect ECR URL, which we normally construct. 

Another thing that would help us understand is the redacted output of `copilot svc package --output-dir infrastructure`? It will generate two files, `$svc-$env.params.json` and `$svc-env.stack.yaml`. The params file output will be most helpful.

**Comment by MatteoGioioso** [comment]
@bvtujo hello, yes give me sometimes. 
Just a quick question that came to my mind: is it a must to run `copilot deploy` before any pipeline?
In my case after `copilot init` I did not proceed with the deployment as I had to run a migration before it, otherwise my task would fail and keep restarting infinitely.

**Comment by MatteoGioioso** [comment]
Also because of that error I needed to set manually the desired count to 0, this completely broke cloudformation deployment and now I cannot recover from it.

**Comment by MatteoGioioso** [comment]
Ok, I have solved it, by deleting everything and move my addons outside copilot, unfortunately I needed to perform a migration in order to make my application run successfully and at the moment that's not possible with copiilot.

After the first successful service deployment I have created a new pipeline and then updated it. This time it worked as expected. 

One more thing which I cannot understand, how can I get the environment (`COPILOT_ENVIRONMENT_NAME`) within the codebuild project? I have run `printenv`, but nothing

**Comment by efekarakus** [comment]
> Just a quick question that came to my mind: is it a must to run `copilot deploy` before any pipeline?

Nope, that shouldn't be necessary! Do you remember if the environment was created with `copilot env init` before creating the pipeline? The commands that are necessary before creating the pipeline are: `copilot app init`, `copilot env init`, and `copilot svc init`. 

> One more thing which I cannot understand, how can I get the environment (COPILOT_ENVIRONMENT_NAME) within the codebuild project? I have run printenv, but nothing

Out of curiosity what's the use case for wanting to get the environment name? in the buildspec we have the `envs` variable that contains the name of the environments:
https://github.com/aws/copilot-cli/blob/5274b273f5f5c554c529e43e3d1a12fc4519051a/templates/cicd/buildspec.yml#L42

**Comment by MatteoGioioso** [comment]
> Nope, that shouldn't be necessary! Do you remember if the environment was created with copilot env init before creating the pipeline? The commands that are necessary before creating the pipeline are: copilot app init, copilot env init, and copilot svc init.

I remember I did `copilot init` and when he asked me to deploy I have declined. I knew already that my deployment would fail as I needed to run a migration first.

> Out of curiosity what's the use case for wanting to get the environment name? in the buildspec we have the envs variable that contains the name of the environments:

My use case would be to run a migration per environment, ex:

```
build:
    commands:
      - echo "Perform migration"
      - |
        	AWS_PROFILE=my-profile ./copilot-linux task run --follow \
		--app my-app \
		--dockerfile db/migrator.Dockerfile \
		--env ${ENV} \
		--task-role my-role-${ENV}
```

But reading through the buildspec code seems like you are building and packaging all the environments at once, so in theory I should loop through the environments and run both migration in the same step, am I right?

**Comment by efekarakus** [comment]
> I remember I did copilot init and when he asked me to deploy I have declined. I knew already that my deployment would fail as I needed to run a migration first.

Ah okay, I think this is why the build stage failed. When Copilot prompts to deploy the first time it will create a "test" environment which is when we create an ECR repository. Since Copilot didn't create an environment, the error above occured.

>  so in theory I should loop through the environments and run both migration in the same step, am I right?

That's right, that sounds good to me!

**Comment by amucunguzi** [comment]
The CodeBuild stage isn't generating the `ContainerImage` parameter in the params output file. NOTE: I only encountered this issue after switching from copilot v1.13.0 to copilot v1.28.0 (the older version was having another issue with deprecated nodejs version for creating Lambda functions).

I have almost the same [buildspec.yml](https://pastebin.com/CMXx9apS) file, [pipeline.yml](https://pastebin.com/xAAdanyh), and [manifest.yml](https://pastebin.com/NtZkuuzD) as the OG (see attached links on Pastebin), and I am using copilot `v1.28.0`. As mentioned, the ContainerImage parameters isn't being generated in the CodeBuild step (see output [JSON here](https://pastebin.com/BPj17rTi)). I have checked the CloudFormation stack template that I am using (2010-09-09), and it does have the ContainerImage parameter defined. 

Because of the missing ContainerImage, the first two lines (from buildspec.yml) below are causing an error:

```
repo=$(cat $CODEBUILD_SRC_DIR/infrastructure/$workload-$env.params.json | jq -r '.Parameters.ContainerImage');
region=$(echo $repo | cut -d'.' -f4);
docker login -u AWS -p $(aws ecr get-login-password --region $region) $AWS_ACCOUNT_ID.dkr.ecr.$region.amazonaws.com
docker tag $image_id $repo;
docker push $repo;
```

For additional details, I'm using the ARM64 version (copilot-linux-arm64-v1.28.0). Also, deploying the apps with "copilot svc deploy" works perfectly fine on staging and production.

Any tips or guidance would be highly appreciated. cc. @bvtujo @efekarakus
